src/Lecture2.hs

```diff
@@ -74,7 +78,11 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt x list =
+  if length list <= x || x < 0
+    then (Nothing , list)
+    else (Just(list!!x), take x list ++ drop (x + 1) list)
```

Comment:
That looks correct! For extra points and bonus challenge, you can try to implement this function using only a single traversal of the list 🧩 

`length` is an expensive operation on lists because it traverses the entire list to calculate its length.

You don't need to check if the index is out of bounds separately. You can use `splitAt` and pattern-matching on the second element of the list. If the index is greater than the list length then `splitAt` returns the entire list as its first element of the pair and an empty list in the second place.

Otherwise, the head of the second list is exactly the element under the desired index.

`length` on lists requires full traversal of the list so there's no way to check if the index is out-of-bounds without traversing the list.

There exist a package that provides the list type that stores its length alongside the list but it's an external dependency:

* https://hackage.haskell.org/package/slist